## Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
from pymystem3 import Mystem

borrowers = pd.read_csv('/datasets/data.csv')
borrowers.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [2]:
borrowers.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 11.3 MB


In [3]:
borrowers.select_dtypes('object').nunique()

education        15
family_status     5
gender            3
income_type       8
purpose          38
dtype: int64

In [4]:
borrowers['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [5]:
borrowers['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

In [6]:
borrowers.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

In [7]:
borrowers['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

**Вывод**

Среди значений days_employed есть отрицательные, что странно, ведь трудовой стаж в днях не может быть со знаком минус. Также трудовой стаж в днях не должен иметь дробную часть. Кроме того, есть запредельные значения в строке 4: 340266.072047. Если разделить это значение на 365 дней в году - получится, что некий пенсионер проработал 932 года! Вот это воля или нездоровый трудоголизм, а может проклятье кредиторов!

Некоторые строковые значения в колонке об образовании различаются регистром. 

Колонки family_status и family_status_id отражают одну и ту же информацию - значит одину колонку можно убрать. Скорее всего то же касается и колонок education и education_id, но это еще предстоит проверить после приведения строковых значений к единому формату.

В выборке есть пропуски в колонках days_employed и total_income, т.е. трудовой стаж и доходы отсутвуют - может быть это дети хотят взять кредит, а может тунеядцы, а может кот-то скрывается от налоговой или забыли заполнить. Всего 21525 - 19351 = 2174 пропущенных значений.

В колонке purpose одинаковые по смыслу цели указаны разными словами фразами - нужно привести к одному формату.

В колонке dob_years есть клиенты с нулевым возрастом, заметно из команды describe в минимальном значении. Т.е. это пропуски.

## Предобработка данных

### Обработка пропусков

In [8]:
missed_count = (borrowers['days_employed'].isnull() & borrowers['total_income'].isnull()).sum()
print(
    "Количество строк, в которых одновременно отсутствуют оба значения 'days_employed' и 'total_income':", 
      missed_count
)
print("Доля пропусков составляет: {:.0%}".format(missed_count / 21525))

Количество строк, в которых одновременно отсутствуют оба значения 'days_employed' и 'total_income': 2174
Доля пропусков составляет: 10%


Во всех строках, в которых пропущено значение трудового стажа в днях, одновременно отсутвует и ежемесячный доход. Точные значения без повторного анкетирования не восстановить (если только они не пропущенны банком, что вряд ли). Так как в обеих колонках - количественные признаки, заменим пропуски *характерными значениями*. Пусть этим значением будет оба раза медиана потому, что в обоих случаях могут быть сильно выделяющиеся значения. Но предварительно приведем значения к единому формату.

Сначала сделаем положительными все значения в колонке 'dayes_emplyed'; и заодно в колонке 'children', так как про количество отрицательных детей в семье не упоминалось.

In [9]:
borrowers['children'] = borrowers['children'].abs()
borrowers['days_employed'] = borrowers['days_employed'].abs()

Предполагаю, что большие значения в колонке days_employed - это не дни, а часы работы. Если так, то поделив выскочек на 24, должны получить разумный временной диапозон по всем значениям. Проверим это 

In [10]:
borrowers.loc[borrowers['days_employed'] > 50000, 'days_employed'] //= 24
print(
    'Минимальный трудовой стаж в годах: {:.0f}'.format(borrowers['days_employed'].min() // 365),
    'Максимальный трудовой стаж в годах: {:.0f}'.format(borrowers['days_employed'].max() // 365), 
    sep='\n'
)

Минимальный трудовой стаж в годах: 0
Максимальный трудовой стаж в годах: 50


Теперь заменим медианой пропущенные значения в количественных признаках 'days_employed' и 'totafl_income'. Количественные, т.к. операции над ними и отношения порядка имеют смысл.

Посмотрим на медианы, сгрупирвав клиентов по типу занятости.

In [11]:
na_indices = borrowers['days_employed'].isna()
borrowers.groupby('income_type')[['days_employed', 'total_income']].median()

days_employed   total_income
income_type                                  
безработный       15266.500000  131339.751676
в декрете          3296.759962   53829.130729
госслужащий        2689.368353  150447.935283
компаньон          1547.382223  172357.950966
пенсионер         15217.000000  118514.486412
предприниматель     520.848083  499163.144947
сотрудник          1574.202821  142594.396847
студент             578.751554   98201.625314

In [12]:
medians = borrowers.groupby('income_type')[['days_employed', 'total_income']].transform('median')
borrowers = borrowers.fillna(medians)
borrowers.loc[na_indices].head(5)

children  days_employed  dob_years education  education_id  \
12         0   15217.000000         65   среднее             1   
26         0    2689.368353         41   среднее             1   
29         0   15217.000000         63   среднее             1   
41         0    2689.368353         50   среднее             1   
55         0   15217.000000         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

     total_income                           purpose  
12  118514.486412                   сыграть свадьбу  
26  150447.935283                       образование  
29  118514.486412  строительство жилой недвижимости  
41  150447.935283  сделка с подержанным автомобилем  
55  118514.486412                   сыграть свадьбу

Успешно заменили отсутвующие значения в колонках 'days_employes' и 'total_income' в данных на медийные значения, учитвыя принадлежность записи к типу занятости.

Заменим отсутвующее значение 'XNA' в колонке 'gender' на самое популярное: 'F'.

In [13]:
borrowers.loc[borrowers['gender'] == 'XNA', 'gender'] = 'F'

Заметим, что в колонке 'children' есть сильно выдающиеся значения: 20 детей. Их мало и они могут сильно искажать характеристики выборки (например выборочное среднее)

In [14]:
borrowers['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

Заменим это значение на моду.

In [15]:
borrowers.loc[borrowers['children'] == 20, 'children'] = 0

In [16]:
borrowers['children'].value_counts()

0    14225
1     4865
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

Осталось заменить нулевые значения в колонке 'dob_years'. Посмотрим их местоположение сгруппировав по типу занятости 'income_type'

In [17]:
table = borrowers.groupby('income_type')['dob_years'].aggregate(['count', 'min', 'median'])
table

count  min  median
income_type                        
безработный          2   31    38.0
в декрете            1   39    39.0
госслужащий       1459    0    40.0
компаньон         5085    0    39.0
пенсионер         3856    0    60.0
предприниматель      2   27    42.5
сотрудник        11119    0    39.0
студент              1   22    22.0

Нулевые значения возраста содержатся в довольно масштабных группах с разными значениями медиан. И медийный пенсионер разительно отличается от остальных средних. Поэтому заполним нулевые значения по группам

In [18]:
zero_people_medians = table.loc[table['min'] == 0, 'median']

for type_name in zero_people_medians.index:
    condition = (borrowers['income_type'] == type_name) & (borrowers['dob_years'] == 0)
    borrowers.loc[condition, 'dob_years'] = zero_people_medians[type_name]

print('Людей с нулевым возрастом осталось:', (borrowers['dob_years'] == 0).sum())

Людей с нулевым возрастом осталось: 0


**Вывод**

- В колонках 'days_employed' и 'total_income' содержатся количественные значения. Пропуски я заменил медианами в каждой колонке соответсвенно.
- В колонке 'children' значения сделал неотрицательными.
- В колонке 'dob_years' средние значения в группах по типу занятости отличаются. Поэтому я заменил нулевые значения  медианами по каждому типу занятости соответсвенно

### Замена типа данных

In [19]:
borrowers = borrowers.astype(
    dtype={'children': 'uint8',
           'days_employed': 'uint32',
           'dob_years': 'uint8',
           'education_id': 'uint8', 
           'family_status_id': 'uint8',
           'total_income': 'uint',
           'debt': 'bool'}
)

borrowers.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null uint8
days_employed       21525 non-null uint32
dob_years           21525 non-null uint8
education           21525 non-null object
education_id        21525 non-null uint8
family_status       21525 non-null object
family_status_id    21525 non-null uint8
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null bool
total_income        21525 non-null uint64
purpose             21525 non-null object
dtypes: bool(1), object(5), uint32(1), uint64(1), uint8(4)
memory usage: 12.7 MB


In [20]:
borrowers.head(5)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   

      family_status  family_status_id gender income_type   debt  total_income  \
0   женат / замужем                 0      F   сотрудник  False        253875   
1   женат / замужем                 0      F   сотрудник  False        112080   
2   женат / замужем                 0      M   сотрудник  False        145885   
3   женат / замужем                 0      M   сотрудник  False        267628   
4  гражданский брак                 1      F   пенсионер  False        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

**Вывод**

- В колонке 'children' заменен тип 'int64' на 'uint8' так как количество детей должно быть положительным, а uint8 позволяет хранить до 255 детей)) Google выдал 87 детей у Федора Васильева
- В колонке 'days_employed' заменен тип 'float64' на 'uint32', так как трудовой стаж в днях во первых неотрицательный (а кто работал -10 дней?), а во вторых количество дней определяется целыми числами. 
- В колонках 'dob_years', 'education_id', 'family_status_id' заменен тип 'int64' на 'uint8', так как это неотрицательные по смыслу значения, а возрастом 200+ лет я не знаю долгожителей...
- В колонке 'debt' заменен тип 'int64' на 'bool', так как этот признак принимает всего два значения была задолженность или ее не было, значит это логический признак

### Обработка дубликатов

Сначала приведем строковые значения в колонке 'education' к единому формату - нижнему регистру. И посмотрим, что получилось.

In [21]:
borrowers['education'] = borrowers['education'].str.lower()
borrowers.pivot_table(index=['education', 'education_id'], values='gender', aggfunc='count')

gender
education           education_id        
высшее              0               5260
начальное           3                282
неоконченное высшее 2                744
среднее             1              15233
ученая степень      4                  6

Итак мы получили пять различных уровней образования, как и в колонке 'education_id'. Сосчитать количество записей можно было не только по признаку 'gender', а по любому другому.

Теперь посмотрим есть ли дублирующиеся строки и удалим их.

In [22]:
print("Количество дубликатов: {}".format(borrowers.duplicated().sum()))

Количество дубликатов: 71


In [23]:
borrowers = borrowers.drop_duplicates().reset_index(drop=True)
print("Количество дубликатов: {}".format(borrowers.duplicated().sum()))

Количество дубликатов: 0


**Вывод**

После приведения значений в колонке 'education' к единому формату, были удалены одинаковые строки методом drop_duplicates. Так как в колонке 'purpose' остались одинаковые по смыслу, но по-разному написанные цели все еще есть риск наличия неочевидных дубликатов. Поэтому нужно повторить тест на наличие дубликатов после форматирования колонки 'purpose'.

### Лемматизация

Посмотрим какие значения содержатся в колонке 'purpose'

In [24]:
borrowers['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Методом пристального вглядывания я обнаружил 5 категорий целей. Они встречаются как отдельные слова (понимаю, что такой метод не всегда сработает). Выделенные категории: ['свадьба', 'жилье', 'недвижимость', 'автомобиль', 'образование']. Заменим цели в колонке 'purpose' на соосветсвующие значения категорий. Для этого напишем функцию и применим ее ко всей колонке с помощью функции apply (pandas.Series.apply обрабатывает значения поэлементно)

In [25]:
m = Mystem()
purpose_types = ['свадьба', 'жилье', 'недвижимость', 'автомобиль', 'образование']

def purpose_mapper(target, map_list):
    for category in map_list:
        if category in m.lemmatize(target):
            return category

Функция готова, теперь проверим ее работу на уникальных значениях колонки 'purpose' - их немного.

In [26]:
for value in borrowers['purpose'].value_counts().index:
    print("Стало: {: >12} | А было: {:<} ".format(purpose_mapper(value, purpose_types), value))

Стало:      свадьба | А было: свадьба 
Стало:      свадьба | А было: на проведение свадьбы 
Стало:      свадьба | А было: сыграть свадьбу 
Стало: недвижимость | А было: операции с недвижимостью 
Стало: недвижимость | А было: покупка коммерческой недвижимости 
Стало:        жилье | А было: операции с жильем 
Стало:        жилье | А было: покупка жилья для сдачи 
Стало: недвижимость | А было: операции с коммерческой недвижимостью 
Стало:        жилье | А было: жилье 
Стало:        жилье | А было: покупка жилья 
Стало:        жилье | А было: покупка жилья для семьи 
Стало: недвижимость | А было: строительство собственной недвижимости 
Стало: недвижимость | А было: недвижимость 
Стало: недвижимость | А было: операции со своей недвижимостью 
Стало: недвижимость | А было: строительство жилой недвижимости 
Стало: недвижимость | А было: покупка недвижимости 
Стало:        жилье | А было: покупка своего жилья 
Стало: недвижимость | А было: строительство недвижимости 
Стало:        жилье | А был

Выглядит прекрасно - справился. Даже 'покупка жилой недвижимости' функция распознала как 'недвижимость', а не 'жилье'. Если бы я использовал *стэммер*, в такой ситуации могла быть ошибка. Теперь применим функцию ко всей колонке.

In [27]:
borrowers['purpose'] = borrowers['purpose'].apply(func=purpose_mapper, args=(purpose_types,))

Проверим еще раз данные на наличие дубликатов и удалим повторы

In [28]:
print("Количество дубликатов: {}".format(borrowers.duplicated().sum()))
borrowers = borrowers.drop_duplicates().reset_index(drop=True)
print("Количество дубликатов: {}".format(borrowers.duplicated().sum()))

Количество дубликатов: 248
Количество дубликатов: 0


**Вывод**

В колонке 'purpose' было много семантически схожих фраз. Было выделено 5 общих категорий для целей получения кредита. Применив лемматизацию к значениям целей в колонке 'purpose' получилось привести все записи к единому формату. После преобразования обнарулось еще 248 дубликатов, которые были удалены.

### Категоризация данных

In [29]:
borrowers['salary_group'] = pd.qcut(borrowers['total_income'], 
                                    q=3, 
                                    labels=['бедный', 'средний', 'богатый'], 
                                    precision=0)

In [30]:
borrowers['children_group'] = pd.cut(borrowers['children'], 
                                   bins=[-1,0,2,20], 
                                   labels=['бездетный', 'малодетный', 'многодетный'])

In [31]:
borrowers_education_dict = borrowers[['education_id', 'education']].drop_duplicates().reset_index(drop=True)
borrowers_education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [32]:
borrowers_fam_status = borrowers[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop=True)
borrowers_fam_status

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

In [33]:
borrowers_categorized = borrowers.drop(columns=['education', 'family_status'])
borrowers_categorized.head(5)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          14177         53             1                 1      F   

  income_type   debt  total_income      purpose salary_group children_group  
0   сотрудник  False        253875        жилье      богатый     малодетный  
1   сотрудник  False        112080   автомобиль       бедный     малодетный  
2   сотрудник  False        145885        жилье      средний      бездетный  
3   сотрудник  False        267628  образование      богатый    многодетный  
4   пенсионер  False        158616      свадьба      средний      бездетный

**Вывод**

В данных можно выделить два словаря для столбцов 'education' и 'family_status', так как соседние колонки 'education_id' и 'family_status_id' - цифровой код того же значения. Таким образом эти колонки из данных можно вынести, удалив при этом много строковых значений, что при большом объеме данных экономит место.

## Ответьте на вопросы

Предполагаю, что для ответов на вопросы нужно сгруппировать данные (возможно результат предыдущего задания здесь должен чем то помочь), сагрегировать их и увидеть есть ли зависимость. Думаю, что можно использовать агрегатную функцию mean к значениям 'debt' по группам. mean() будет равна доле единичных значений в группе, кроме того что это среднее значение. Так можно посчитать долю нарушителей в группах

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [34]:
borrowers.pivot_table(index=['children_group'], values=['debt'], aggfunc=['count', 'sum', 'mean'])

count     sum      mean
                 debt    debt      debt
children_group                         
бездетный       13964  1070.0  0.076626
малодетный       6863   639.0  0.093108
многодетный       379    31.0  0.081794

**Вывод**

9 девять человек с пятью детьми - самые белые и пушистые и отдают кредиты в срок.
Самая низкая доля нарушителей среди подавляющего бездетного большенства. Можно проверить гипотезу "О святости 5-детных", но я зависимости не уидел, значения довольно близки и некоторые выборки слишком малы чтобы давать оценку.


- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [35]:
borrowers.pivot_table(index=['family_status'], values=['debt'], aggfunc=['count', 'sum', 'mean'])

count    sum      mean
                        debt   debt      debt
family_status                                
Не женат / не замужем   2792  274.0  0.098138
в разводе               1193   85.0  0.071249
вдовец / вдова           946   63.0  0.066596
гражданский брак        4129  388.0  0.093969
женат / замужем        12146  930.0  0.076568

**Вывод**

Хотя отличие и небольшое может вдовцы/вдовы чаще других категорий исполняют обязательства по кредитам.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [36]:
borrowers.pivot_table(index=['salary_group'], values=['debt'], aggfunc=['count', 'sum', 'mean'])

count    sum      mean
              debt   debt      debt
salary_group                       
бедный        7069  581.0  0.082190
средний       7120  631.0  0.088624
богатый       7017  528.0  0.075246

**Вывод**

- Предприниматели не оставляют долгов. 
- Среди госслужащих и пенсионеров редко встречаются недобросовестные заемщики
- Чуть хуже ситуация у сотрудников.
- Хуже всех ведут себя безработные и в декрете.



- Как разные цели кредита влияют на его возврат в срок?

In [37]:
borrowers.pivot_table(index=['purpose'], values=['debt'], aggfunc=['count', 'sum', 'mean'])

count    sum      mean
              debt   debt      debt
purpose                            
автомобиль    4272  402.0  0.094101
жилье         4406  308.0  0.069905
недвижимость  6258  474.0  0.075743
образование   3964  370.0  0.093340
свадьба       2306  186.0  0.080659

**Вывод**

Чаще всего исполняются обязательства по кредитам на жилье, видимо потому, что жилье находится в залоге у кредитора.

## Общий вывод

Думаю, что существует зависимость между типом занятости (а значит и уровнем дохода, т.к. в каждом типе занятости средний доход разный) и исполнением обязательств по кредитам. Чем выше доход тем выше вероятность, что заемщик вернет деньги вовремя

Все рассмотренные признаки: уровень дохода, цели кредита, количество детей, семейное положение так или иначе влияют на степень доверия к клиенту и его исполнительность по своевременному возврату заемных средств. В целом можно сказать, что люди у которых оформлено семейное положение, еще нет детей или их много; из группы 'богатых' и хотят купить недвижимость или квартиру скорее вернут заем в срок. Может это как то связано со словом ответсвенность и те кто справляется с большим количеством последствий - более надежные клиенты</p>

Во всяком случае для принятия решения по каждому клиенту нужно учитывать каждый из рассмотренных признаков.